File: projects/client/src/networks/carbon/FormulaCompiler/lexer.ts
import { SyntaxError } from './helpers';
import {
	FunctionHandleToken,
	NumberToken,
	StreamToken,
	Token,
	TokenCharacters,
	TokenType,
	getTokenTypeFromCharacter,
} from './tokens';

// Type Checking Functions
const isNumber = (char: string) => '0123456789'.includes(char) || char === '.';
const isWhitespace = (char: string) => ' \t\n'.includes(char);
const isLetter = (char: string) => /[a-zAZ]/i.test(char);
const isFunctionHandle = (char: string) => char === '@';
const isStream = isLetter;
const isOperand = (char: string) => Object.values(TokenCharacters).includes(char);
export class Lexer {
	currentIndex: number;
	currentCharacter: string | null;

	constructor(public text: string) {
		this.text = text;
		this.currentIndex = -1;
		this.currentCharacter = '';
		this.advance();
	}

	advance() {
		this.currentIndex++;
		if (this.currentIndex < this.text.length) {
			this.currentCharacter = this.text[this.currentIndex];
		} else {
			this.currentCharacter = null;
		}
	}

	generateTokens() {
		const ret: Token[] = [];

		while (this.currentCharacter !== null) {
			if (isWhitespace(this.currentCharacter)) {
				this.advance();
			} else if (isNumber(this.currentCharacter)) {
				ret.push(this.generateNumber());
			} else if (isFunctionHandle(this.currentCharacter)) {
				ret.push(this.generateFunctionHandle());
			} else if (isStream(this.currentCharacter)) {
				ret.push(this.generateStream());
			} else if (isOperand(this.currentCharacter)) {
				ret.push(this.generateOperand());
			} else {
				throw new SyntaxError(`Invalid character: ${this.currentCharacter}`, {
					startIndex: this.currentIndex,
					endIndex: this.currentIndex + 1,
					value: this.currentCharacter,
				});
			}
		}
		return ret;
	}

	generateNumber(): NumberToken {
		const startIndex = this.currentIndex;
		let numberString = this.currentCharacter;
		if (numberString === null)
			throw new SyntaxError(undefined, {
				startIndex: this.currentIndex,
				endIndex: this.currentIndex + 1,
				value: '',
			});
		this.advance();

		while (this.currentCharacter !== null && isNumber(this.currentCharacter)) {
			numberString += this.currentCharacter;
			this.advance();
		}

		return new NumberToken({
			startIndex,
			endIndex: this.currentIndex,
			tokenType: TokenType.NUMBER,
			value: numberString,
		});
	}

	generateFunctionHandle(): FunctionHandleToken {
		const startIndex = this.currentIndex;
		let functionHandleString = this.currentCharacter;
		if (functionHandleString === null)
			throw new SyntaxError(undefined, {
				startIndex: this.currentIndex,
				endIndex: this.currentIndex + 1,
				value: '',
			});
		this.advance();

		while (this.currentCharacter !== null && isLetter(this.currentCharacter)) {
			functionHandleString += this.currentCharacter;
			this.advance();
		}

		return new FunctionHandleToken({
			startIndex,
			endIndex: this.currentIndex,
			tokenType: TokenType.FUNCTION_HANDLE,
			value: functionHandleString,
		});
	}

	generateStream(): StreamToken {
		const startIndex = this.currentIndex;
		let streamString = this.currentCharacter;
		if (streamString === null)
			throw new SyntaxError(undefined, {
				startIndex: this.currentIndex,
				endIndex: this.currentIndex + 1,
				value: '',
			});
		this.advance();

		while (this.currentCharacter !== null && isStream(this.currentCharacter)) {
			streamString += this.currentCharacter;
			this.advance();
		}

		return new StreamToken({
			startIndex,
			endIndex: this.currentIndex,
			tokenType: TokenType.STREAM,
			value: streamString,
		});
	}

	generateOperand(): Token {
		const startIndex = this.currentIndex;
		const operandString = this.currentCharacter;
		if (operandString === null)
			throw new SyntaxError(undefined, {
				startIndex: this.currentIndex,
				endIndex: this.currentIndex + 1,
				value: '',
			});
		this.advance();

		return new Token({
			startIndex,
			endIndex: this.currentIndex,
			tokenType: getTokenTypeFromCharacter(operandString) as TokenType,
		});
	}
}

File: packages/python/combocurve/combocurve/science/network_module/parser/lexer.py
import re
from combocurve.science.network_module.parser.tokens import (Token, TokenType, STR_TO_TOKEN_MAP, StreamToken,
                                                             NumberToken, FunctionHandleToken)

WHITESPACE = ' \n\t'
DIGITS = '0123456789'
letter_re = re.compile('[a-zA-Z]')


def match_whitespace(char: str):
    return char in WHITESPACE


def match_number(char: str):
    return char in (DIGITS + '.')


def match_stream(char: str):
    return letter_re.fullmatch(char)


def match_function_start(char: str):
    return char == '@'


def match_operand(char: str):
    return char in STR_TO_TOKEN_MAP


class Lexer:
    '''
    Lexer class will take raw string input and give tokens output
    '''
    def __init__(self, text: str):
        self.text: str = text
        self.current_index: int = -1
        self.current_char: None | str = None
        self.advance()

    def advance(self):
        self.current_index += 1
        if self.current_index < len(self.text):
            self.current_char = self.text[self.current_index]
        else:
            self.current_char = None

    def generate_tokens(self):
        ret: list[Token] = []
        while self.current_char is not None:
            if match_whitespace(self.current_char):
                self.advance()
            elif match_number(self.current_char):
                ret += [self.generate_number()]
            elif match_function_start(self.current_char):
                ret += [self.generate_function_handle()]
            elif match_stream(self.current_char):
                ret += [self.generate_stream()]
            elif match_operand(self.current_char):
                ret += [self.generate_operand()]
            else:
                raise ValueError(f"Illegal character '{self.current_char}'")

        return ret

    def generate_operand(self) -> Token:
        start_index: int = self.current_index

        operand_str: str = self.current_char
        self.advance()
        return Token(start_index, self.current_index, STR_TO_TOKEN_MAP[operand_str])

    def generate_stream(self) -> StreamToken:
        start_index: int = self.current_index
        stream_str: str = self.current_char
        self.advance()
        while self.current_char is not None and letter_re.fullmatch(self.current_char):
            stream_str += self.current_char
            self.advance()

        return StreamToken(start_index, self.current_index, TokenType.STREAM, stream_str)

    def generate_function_handle(self):
        start_index: int = self.current_index
        function_handle_str: str = self.current_char

        self.advance()
        while self.current_char is not None and letter_re.fullmatch(self.current_char):
            function_handle_str += self.current_char
            self.advance()

        return FunctionHandleToken(start_index, self.current_index, TokenType.FUNCTION_HANDLE, function_handle_str)

    def generate_number(self):
        start_index: int = self.current_index
        number_str: str = self.current_char
        self.advance()

        while self.current_char is not None and match_number(self.current_char):
            number_str += self.current_char
            self.advance()

        ## TODO: delete these lines, python float() can work with ".3" and "3."
        ## do not ned to convert them to a more human-read format
        # if number_str.startswith('.'):
        #     number_str = '0' + number_str

        # if number_str.endswith('.'):
        #     number_str += '0'

        return NumberToken(start_index, self.current_index, TokenType.NUMBER, number_str)
